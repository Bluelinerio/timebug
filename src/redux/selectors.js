// @flow
import { getUserState, getCms, getStorageLoaded } from './rootReducer';
import { UNDETERMINED, ANONYMOUS, AUTHENTICATING } from '../services/apollo/models';
import type { Progress } from '../services/apollo/models';
import type { Colors, Step } from '../services/cms';
import type { StepsState } from './reducers/cms.reducer'

const sortSteps = (a: Step, b: Step) => a.number - b.number
const isStorageLoaded = (state: any):boolean => getStorageLoaded(state);
const steps = (state) => getCms(state).steps
const sortedSteps = (state: any) :[Step] => Object.values( steps(state) ).sort(sortSteps) 
const stepColors = (state: any):Colors => getCms(state).colors.steps;
const isCMSLoading = (state: any) => getCms(state).requestCount > 0
const totalNumberOfSteps = (state: any) => getCms(state).totalNumberOfSteps
const colors = (state:any) => getCms(state).colors
const aboutText = (state:any) => getCms(state).about

const user = (state: any): ?User =>
	typeof getUserState(state) === 'string' ? null : getUserState(state)
const isUserStateUNDETERMINED = (state: any): boolean => getUserState(state) === UNDETERMINED
const isUserStateAUTHENTICATING = (state: any): boolean => getUserState(state) === AUTHENTICATING
const isLoggedIn = (state: any) : boolean => !!user(state)

const progress = (state: any): ?Progress => user(state) ? user(state).progress : null;
const currentStepNumber = (state: any): number => progress(state) ? progress(state).step : 0;
const currentStep = (state: any): Step => steps(state)[currentStepNumber(state)];
const currentStepColor = (state: any) => stepColors(state)[currentStepNumber(state)];
const assignments = (state: any) => currentStep(state).refAssignment.map(i => i.fields);
const colorForStep = (step: number) => (state:any) => stepColors(state)[step]
const step = (number: number) => (state:any) => steps(state)[number]

export default {
	isStorageLoaded,
	getCms,
	sortedSteps,
	colorForStep,
	aboutText,
	step,
	colors,
	isCMSLoading,
	totalNumberOfSteps,
	currentStepNumber,
	currentStep,
	progress,
	assignments,
	currentStepColor,
	user,
	isLoggedIn,
	isUserStateUNDETERMINED,
	isUserStateAUTHENTICATING
}