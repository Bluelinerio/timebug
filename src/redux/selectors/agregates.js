import R from 'ramda';
import { NavigationReducerKeys } from '../reducers/agregates.navigation.reducer';
import { AgregateReducerKeys } from '../reducers/agregates.reducer';

export const stepGuideLens = R.lensPath([
  AgregateReducerKeys.agregate,
  NavigationReducerKeys.pageVisits,
  NavigationReducerKeys.pages.stepGuide,
]);
export const stepWorkbookLens = R.lensPath([
  AgregateReducerKeys.agregate,
  NavigationReducerKeys.pageVisits,
  NavigationReducerKeys.pages.stepWorkbook,
]);

export const sortPageViewByKey = prop =>
  R.sort((a, b) => b[1][prop] - a[1][prop]);
export const sortPageViewByStepID = R.sort((a, b) => a[0] - b[0]);
export const sortPageViewTime = sortPageViewByKey('last');
export const sortPageViewCounts = sortPageViewByKey('count');

export const packPairs = key => array => ({
  [key]: array[0],
  ...array[1],
});

export const stepGuideSortedByTime = R.compose(
  R.map(packPairs('stepId')),
  sortPageViewTime,
  R.toPairs,
  R.view(stepGuideLens)
);

export const stepGuideSortedByCounts = R.compose(
  R.map(packPairs('stepId')),
  sortPageViewCounts,
  R.toPairs,
  R.view(stepGuideLens)
);

export const lastStepGuideVisited = R.compose(R.nth(0), stepGuideSortedByTime);
export const mostFrequentStepGuideVisited = R.compose(
  R.nth(0),
  stepGuideSortedByCounts
);

export const stepWorkbookSortedByTime = R.compose(
  R.map(packPairs('stepId')),
  sortPageViewTime,
  R.toPairs,
  R.view(stepWorkbookLens)
);

export const stepWorkbookSortedByCounts = R.compose(
  R.map(packPairs('stepId')),
  sortPageViewCounts,
  R.toPairs,
  R.view(stepWorkbookLens)
);

export const lastStepWorkbookVisited = R.compose(
  R.nth(0),
  stepWorkbookSortedByTime
);
export const mostFrequentStepWorkbookVisited = R.compose(
  R.nth(0),
  stepWorkbookSortedByCounts
);
