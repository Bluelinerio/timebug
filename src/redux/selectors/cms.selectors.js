// @flow
import R from 'ramda'
import { filterKeys, viewOr } from './utils'
import { stepIds } from '../../constants/steps'
import * as reducerLenses from '../lenses/reducer.lenses'

export const filterStepIds = filterKeys(stepIds)

const byNumber = R.ascend(R.prop('number'))

// CMS
export const steps = viewOr({}, reducerLenses.steps)
export const allSteps = R.compose(R.values, steps)
export const sortedSteps = R.pipe(
  steps,
  R.values,
  R.sort(byNumber)
)
export const stepColors = viewOr({}, reducerLenses.stepColors)
export const phaseColors = viewOr({}, reducerLenses.phaseColors)
export const introSlides = viewOr([], reducerLenses.introSlides)
export const isCMSLoading = R.compose(
  i => i > 0,
  viewOr(0, reducerLenses.cmsRequestCount),
  // R.equals(0),
  // R.not
)
export const totalNumberOfSteps = viewOr(30, reducerLenses.totalNumberOfSteps)
export const colors = viewOr(undefined, reducerLenses.colors)
export const uniqueColors = state => R.uniq(
  [...R.values(phaseColors(state)), ...R.values(stepColors(state))]
)
export const meditations = viewOr([], reducerLenses.meditations)
export const step = R.curry((state, stepId) => R.prop(stepId, steps(state)))
export const pages = viewOr([], reducerLenses.pages)
export const appInstructions = viewOr(undefined, reducerLenses.appInstructions)

export default {
  steps,
  allSteps,
  sortedSteps,
  meditations,
  phaseColors,
  stepColors,
  introSlides,
  step,
  pages,
  appInstructions,
  colors,
  uniqueColors,
  isCMSLoading,
  totalNumberOfSteps
}