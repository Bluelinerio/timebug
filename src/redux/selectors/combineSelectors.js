//@flow
export type Selector = (state: {}, ownProps: {}) => *

export const combineSelectors = (selectors: {}) => (state: {}, ownProps: {}): * => {
  if (__DEV__) {
    const keyForNotAFunction = Object.keys(selectors).find(key => typeof selectors[key] !== 'function')
    if(keyForNotAFunction) {
      throw new Error(`combineSelectors error: expected a function ${keyForNotAFunction} got ${selectors[keyForNotAFunction]}`)
    }
  }
  return Object.keys(selectors).reduce(
    (sum, key) => ({
      ...sum,
      [key]: selectors[key](state, ownProps)
    }),
    {}
  )
}
