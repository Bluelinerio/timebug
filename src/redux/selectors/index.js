// @flow
import R from 'ramda';
import {
  getUserState,
  getCms,
  getFormData,
  getAwards,
  getUIState,
  getCheckinState,
  getPersistState,
  getGoals,
} from './rootReducer.selectors';
import {
  // UNDETERMINED,
  ANONYMOUS,
  AUTHENTICATING,
} from '../../services/apollo/models';
// models
import workbooks from '../../screens/WorkbookScreen/forms';
import { removeIvalidValuesInsteadOfDoingAnyMigrationForNow } from '../tcomb';
import type { User, Form } from '../../services/apollo/models';
import type { Step, Slide } from '../../services/cms';
import { getStepColors, getPhaseColors } from '../../services/dummyCms';

export const filterWithKeys = (predicate, obj) =>
  R.pipe(R.toPairs, R.filter(R.apply(predicate)), R.fromPairs)(obj);

function isPositiveInteger(n) {
  return n >>> 0 === parseFloat(n);
}

export const filterNumbers = (obj: {}) =>
  filterWithKeys(key => isPositiveInteger(key), obj);

export const filterKeys = (keys: [string]) => (obj: {}) =>
  filterWithKeys(key => keys.includes(key), obj);
export const range = (start, end) =>
  Array(end - start)
    .fill()
    .map((v, i) => i + start);

export const stepIds = range(1, 30).map((v, i) => i.toString());

export const filterStepIds = filterKeys(stepIds);

// CMS
const sortSteps = (a: Step, b: Step) => a.number - b.number;
const steps = (state: any): [Step] => getCms(state).steps;
const sortedSteps = (state: any): [Step] =>
  Object.values(steps(state)).sort(sortSteps);
const stepColors = (state: any): { number: string } =>
  getCms(state).colors.steps;
const introSlides = (state: any): [Slide] =>
  getCms(state).onboardingPages.intro.slides;
const phaseColors = (state: any): { string: string } =>
  getCms(state).colors.phases;

const isCMSLoading = (state: any) => getCms(state).requestCount > 0;
const totalNumberOfSteps = (state: any) => getCms(state).totalNumberOfSteps;
const colors = (state: any) => getCms(state).colors;
const uniqueColors = (state: any) => [
  ...new Set([
    ...Object.values(getCms(state).colors.steps),
    ...Object.values(getCms(state).colors.phases),
  ]),
];
const meditations = (state: any) => getCms(state).meditations;
const step = (number: number) => (state: any) => steps(state)[number];

const pages = state => getCms(state).pages;
const appInstructions = (state: any) => pages(state)['AppInstructions'];

// User
const getUser = (state: any): ?User => getUserState(state);
const user = (state: any): ?User =>
  typeof getUserState(state) === 'string' ? null : getUserState(state);
const userId = (state: any) => user(state) && user(state).id;

const isNotLoggedIn = (state: any): boolean => !user(state);
const isLoggedIn = (state: any): boolean => !!user(state);
const isAnonymous = (state: any): boolean => getUserState(state) === ANONYMOUS;
const isAuthenticating = (state: any): boolean =>
  getUserState(state) === AUTHENTICATING;

const sortForms = (a: Form, b: Form) => a.stepId - b.stepId;
const sortFormsChronologically = (a: Form, b: Form) =>
  Date.parse(b.updatedAt) - Date.parse(a.updatedAt);
// stepId on the server is an Int!. A clear idea how to

const hasCompletedForms = (state: any): boolean =>
  user(state) && user(state).forms.length > 0;

const hasNoCompletedForms = R.compose(R.not, hasCompletedForms);

const completedForms = (state: any): [Form] =>
  user(state) ? user(state).forms.map(f => f) : [];

const completedFormsData = (state: any) =>
  completedForms(state).reduce(
    (forms, form) => ({
      ...forms,
      [form.stepId]: form.data,
    }),
    {}
  );

const completedFormsChronologically = (state: any): [Form] =>
  completedForms(state).sort(sortFormsChronologically);

const sortedCompletedForms = (state: any): [Form] =>
  completedForms(state).sort(sortForms);

const completedStepIds = (state: any): [string] =>
  completedForms(state).map(f => f.stepId);

const formWithStepId = (state: any) => (stepId: string): Form =>
  completedForms(state).find(f => f.stepId === stepId);

// form data
const formData = (state: any) => getFormData(state).data;
const incompleteFormsData = (state: any) =>
  filterStepIds(getFormData(state).data);

const EDITING_FORMS_SUPPORTTED = false;
const buttonTitleForFormCompletion = ({ completedForms, incomplete }) => {
  if (Object.keys(incomplete).length > 0) {
    return 'Resume';
  } else if (
    EDITING_FORMS_SUPPORTTED &&
    completedForms &&
    Object.keys(completedForms).length > 0
  ) {
    return 'Edit';
  } else {
    return 'Start';
  }
};

const sortedStepsWithForms = state => {
  const completed = completedForms(state);
  const incompleteForms = incompleteFormsData(state);

  const { latestStepId } = Object.keys(incompleteForms).reduce(
    (sum, latestStepId) => {
      const timeStamp =
        incompleteForms[latestStepId] &&
        incompleteForms[latestStepId].timeStamp;
      if (timeStamp) {
        if (!sum.timeStamp || sum.timeStamp < timeStamp) {
          return {
            latestStepId,
            timeStamp,
          };
        }
      }
      return sum;
    },
    {}
  );

  return {
    sortedStepsWithForms: sortedSteps(state)
      .map(step => ({
        completedForms: completed.find(f => f.stepId),
        incomplete: filterNumbers(incompleteForms[step.stepId]),
        step,
      }))
      .map(step => ({
        ...step,
        buttonTitleForFormCompletion: buttonTitleForFormCompletion(step),
      })),
    latestStepId,
  };
};

const buttonTitlesForFormCompletion = state => stepId => {
  return R.compose(
    buttonTitleForFormCompletion,
    R.find(i => i.step.stepId === stepId),
    R.prop('sortedStepsWithForms'),
    sortedStepsWithForms
  )(state);
};

const modelsAndDataForExercise = (state: any) => (stepId: string) => {
  //TComb Forms helpers

  const models = workbooks[stepId];
  const localData = getFormData(state).data[stepId];
  if (!localData) {
    return {
      models,
      formData: {},
    };
  }

  const mergeForm = (sum, formKey) => ({
    ...sum,
    ...(localData[formKey] && {
      [formKey]: removeIvalidValuesInsteadOfDoingAnyMigrationForNow(
        models[formKey],
        localData[formKey]
      ),
    }),
  });
  const formData = Object.keys(models).reduce(mergeForm, {});

  return {
    models,
    formData,
  };
};

const isSynchingFormData = (state: any) => getFormData(state).requestCount > 0;

const loadingFormData = (state: any) => getFormData(state).loadingFormData;

const isStepCompleted = (state: any) => {
  const _stepCompleted = (step: number) => {
    const currentUser = user(state);
    if (!currentUser) return false;
    const { forms } = currentUser;
    const completed =
      forms && forms.find(form => `${form.stepId}` === `${step}`)
        ? true
        : false;
    return completed;
  };
  return _stepCompleted;
};
/**
 * Should check CMS for colors, for now it's hardcoded
 */
const statefullStepColors = (state: any) => getStepColors(state);
const overridePhaseColors = (state: any) => getPhaseColors(state);
/**
 * Award
 */

const awardModelsAndData = (state: any) => getAwards(state);

const awardModelForStep = (state: any) => (step: number) =>
  awardModelsAndData(state).models[step]
    ? awardModelsAndData(state).models[step]
    : {};

const awardDataForStep = (state: any) => (step: number) =>
  awardModelsAndData(state).data[step]
    ? awardModelsAndData(state).data[step]
    : {};

const awardModelAndDataForStep = (state: any) => (step: number) => {
  const data = awardDataForStep(state)(step);
  const model = awardModelForStep(state)(step);
  return {
    data,
    model,
  };
};

/**
 * UI
 */

const stateForScreen = (state: any) => (screen: string) =>
  getUIState(state).screens[screen] || {};

/**
 * Checkins
 */
const getCheckins = (state: any) => getCheckinState(state).checkins;

/**
 * Persist Status
 */

const hasStoreLoaded = (state: any) =>
  getPersistState(state).status === 'LOADED';

/**
 * Goals
 */
const getGoalsData = (state: any) => getGoals(state).data;

const getGoalsDataForGoal = (state: any) => {
  const data = getGoalsData(state);
  return (goalId: string) => {
    const dataForGoals = data[goalId];
    return dataForGoals;
  };
};

const getGoalStepsForGoalAndForm = (state: any) => {
  const data = getGoalsData(state);
  return (goalId: string, formId: string) => {
    const dataForGoals = data[goalId];
    const form = dataForGoals[formId];
    const onlyValueElements = Object.keys(form)
      .filter(key => key !== 'timestamp' && key !== 'id')
      .reduce((object, key) => {
        return {
          ...object,
          [key]: form[key],
        };
      }, {});
    return onlyValueElements;
  };
};

const getGoalsStepsForGoalAndFormStateless = (
  data: any,
  goalId: string,
  formId: string
) => {
  const dataForGoals = data[goalId] || {};
  const form = dataForGoals[formId] || {};
  const onlyValueElements = Object.keys(form)
    .filter(key => key !== 'timestamp' && key !== 'id')
    .reduce((object, key) => {
      return {
        ...object,
        [key]: form[key],
      };
    }, {});
  return onlyValueElements;
};

const formHelpSlides = state => getCms(state).helpSlides || {};

const selectors = {
  getCms,
  sortedSteps,
  sortedStepsWithForms,
  buttonTitlesForFormCompletion,
  getUser,
  steps,
  meditations,
  phaseColors,
  introSlides,
  step,
  pages,
  appInstructions,
  colors,
  uniqueColors,
  stepColors,
  isCMSLoading,
  totalNumberOfSteps,
  user,
  userId,
  isLoggedIn,
  isNotLoggedIn,
  isAnonymous,
  isAuthenticating,
  hasCompletedForms,
  hasNoCompletedForms,
  completedForms,
  sortedCompletedForms,
  completedFormsChronologically,
  completedFormsData,
  completedStepIds,
  formWithStepId,
  modelsAndDataForExercise,
  formData,
  incompleteFormsData,
  isSynchingFormData,
  loadingFormData,
  awardModelAndDataForStep,
  statefullStepColors,
  overridePhaseColors,
  stateForScreen,
  getCheckins,
  hasStoreLoaded,
  getGoalsData,
  getGoalsDataForGoal,
  getGoalStepsForGoalAndForm,
  getGoalsStepsForGoalAndFormStateless,
  formHelpSlides,
  isStepCompleted,
};

export default selectors;
