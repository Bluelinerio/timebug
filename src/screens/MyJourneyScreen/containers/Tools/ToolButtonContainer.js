//@flow
import { connect }                     from 'react-redux'
import { compose }                     from 'recompose'
import mapNavigationDispatch           from '2020_HOC/NavigationServiceHOC'
import ToolButton, { ToolButtonProps } from '../../components/Tools/ToolButton'
import {
  mapPhaseAndCompletionToKey,
  getTextColorForPhase,
}                                      from '../../utils/colorsForStep'
import { goToTool }                    from '../../../../redux/actions/nav.actions'
import type { GoToToolParams }         from '../../../../redux/actions/nav.actions'

type ToolButtonDispatchProps = {
  goToTool: GoToToolParams => any,
}

type ToolButtonContainerProps = {
  navigation: any,
  user: any,
  stepColors: any,
  tool: any,
  locked?: boolean,
}
const mapDispatchToProps = (dispatch: any): ToolButtonDispatchProps => ({
  goToTool: (params: GoToToolParams) => dispatch(goToTool(params)),
})

//TODO: Get icon from different source, not the steps directly
const merge = (
  stateProps: any,
  dispatchProps: ToolButtonDispatchProps,
  ownProps: ToolButtonContainerProps
): ToolButtonProps => {
  const { stepColors, step, tool, goToTool, locked = false } = ownProps

  const { icon } = step
  const { title, subtitle, content, phase } = tool

  return {
    title,
    subtitle,
    content,
    phase,
    source: icon && icon.uri,
    onPress: goToTool,
    tool,
    containerBackgroundColor: stepColors[mapPhaseAndCompletionToKey(phase)],
    textStyle: getTextColorForPhase(phase),
    locked,
  }
}

export default compose(mapNavigationDispatch(mapDispatchToProps), connect(null, null, merge))(ToolButton)
