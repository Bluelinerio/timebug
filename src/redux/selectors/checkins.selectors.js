// @flow
import R        from 'ramda'
import {
  // CHECKIN_PHYSICALLY,
  // CHECKIN_EMOTIONALLY,
  // CHECKIN_MENTALLY,
  // CHECKIN_DAILY,
  // CHECKIN_MOOD,
  CHECKIN_MEDITATION,
  // CHECKIN_EMOJI,
}               from '../../constants/checkins'
import { user } from './user.selectors'

const checkins                 = R.compose(
  R.prop('checkins'),
  user
)
const isMeditationCheckin      = R.whereEq({ name: CHECKIN_MEDITATION })
const showUserMeditationOption = R.compose(
  R.not(-1),
  R.findIndex(isMeditationCheckin),
  checkins
)

export default {
  checkins,
  isMeditationCheckin,
  showUserMeditationOption
}