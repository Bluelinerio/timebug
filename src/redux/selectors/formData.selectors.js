// @flow
import R from 'ramda'
import workbooks from '../../screens/WorkbookScreen/forms'
import { getFormData } from './rootReducer.selectors'
import { removeIvalidValuesInsteadOfDoingAnyMigrationForNow } from '../tcomb'
import { filterKeys, filterNumbers, viewOr } from './utils'
import { stepIds } from '../../constants/steps'
import formsSelectors from './forms.selectors'
import cmsSelectors from './cms.selectors'
import combineSelectors from './combineSelectors';

const { completedForms } = formsSelectors
const { sortedSteps, allSteps } = cmsSelectors

export const filterStepIds = filterKeys(stepIds)
const EDITING_FORMS_SUPPORTTED = false

const formData = R.compose(
	viewOr({}, R.lensProp('data')),
	getFormData
)

const incompleteFormsData = (state: any) => filterStepIds(getFormData(state).data)

export const incompleteForms = R.compose(
	R.values,
	filterStepIds,
	formData
)

const buttonTitleForFormCompletion = ( { forms: { complete, incomplete } }) => {
	if (incomplete || incomplete.length > 0) {
		return 'Resume'
	} else if (EDITING_FORMS_SUPPORTTED && complete && incomplete) {
		return 'Edit'
	} else {
		return 'Start'
	}
}

// export const completedFormsGroupedByStepIdSortedByDate = R.compose(
// 	R.groupBy(R.prop('stepId')),
// 	completedForms
// )

// R.compose(
// 	R.sortBy(R.prop('updatedAt')),
// 	R.find(f => f.stepId === step.stepId)
// )
// R.compose(
// 	R.sortBy(R.prop('timeStamp')),
// 	R.find(f => f.stepId === step.stepId)
// )

export const combineSteps = R.compose(
	R.sortWith([
		R.descend(R.path(['forms', 'incomplete', 'updatedAt'])),
		R.ascend(R.path(['forms', 'complete', 'timeStamp']))
	]),
	({
		completedForms,
		incompleteForms,
		allSteps,
	}) => allSteps.map(step => ({
			...step,
			forms: ({
				complete: R.find(f => f.stepId === step.stepId, completedForms),
				incomplete: R.find(f => f.stepId === step.stepId, incompleteForms),
			})
		}
	)),
	combineSelectors({
		completedForms,
		incompleteForms,
		allSteps,
	})
)

const withSelectors = selectors => item => ({
	...item,
	...combineSelectors(selectors)(item)
})

export const buttonTitlesForFormCompletion = state => stepId => {
	return R.compose(
		buttonTitleForFormCompletion,
		R.find(i => i.step.stepId === stepId),
		R.prop('sortedStepsWithForms'),
		sortedStepsWithForms
	)(state)
}

export const sortedStepsWithForms = state => {
	const steps = combineSteps(state).map(withSelectors(buttonTitleForFormCompletion));
	const latestStepId = R.view(R.lensPath([0, 'stepId']), steps)
	debugger;
	return {
		steps,
		latestStepId
	}
}


export const modelsAndDataForExercise = (state: any) => (stepId: string) => {
	//TComb Forms helpers

	const models = workbooks[stepId]
	const localData = getFormData(state).data[stepId]

	if (!localData) {
		return {
			models,
			formData: {}
		}
	}

	const mergeForm = (sum, formKey) => ({
		...sum,
		...(localData[formKey] && {
			[formKey]: removeIvalidValuesInsteadOfDoingAnyMigrationForNow(models[formKey], localData[formKey])
		})
	})
	const formData = Object.keys(models).reduce(mergeForm, {})

	return {
		models,
		formData
	}
}

export const isSynchingFormData = (state: any) => getFormData(state).requestCount > 0

export default {
	sortedStepsWithForms,
	buttonTitlesForFormCompletion,
	completedForms,
	modelsAndDataForExercise,
	formData,
	incompleteFormsData,
	incompleteForms,
	isSynchingFormData
}
