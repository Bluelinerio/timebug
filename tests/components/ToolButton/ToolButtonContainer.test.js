import React from 'react'
import { createMockStore } from 'redux-test-utils'
import selectors from '../../../src/redux/selectors'
import shallowWithStore from '../../shallowWithStore'
import ToolButtonContainer from '../../../src/screens/MyJourneyScreen/containers/Tools/ToolButtonContainer'

describe('Connected Tool screen content component', function() {
  it('Should render succesfully the connected component', function() {
    const state = {}
    const store = createMockStore(state)
    const stepColors = selectors.statefullStepColors(state)
    const dummyData = {
      step: {
        number: 1,
        icon: {
          uri:
            'https://images.ctfassets.net/6h184bey8vl3/Rknt0PorAIG4uoMQAg4KM/d697d93618b298ce10d1606c7d3f573a/TimeBug-Icons-_Step-1_.svg',
        },
      },
      tool: {
        title: 'Some title',
        subtitle: 'Some subtitle',
        content: 'Some content',
        phase: 'Some phase',
      },
      stepColors,
    }
    const component = shallowWithStore(
      <ToolButtonContainer {...dummyData} />,
      store
    )
    expect(component).toMatchSnapshot()
  })
})
