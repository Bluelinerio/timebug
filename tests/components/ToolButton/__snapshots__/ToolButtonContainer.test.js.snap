// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connected Tool screen content component Should render succesfully the connected component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(ToolButton)
    step={
      Object {
        "icon": Object {
          "uri": "https://images.ctfassets.net/6h184bey8vl3/Rknt0PorAIG4uoMQAg4KM/d697d93618b298ce10d1606c7d3f573a/TimeBug-Icons-_Step-1_.svg",
        },
        "number": 1,
      }
    }
    stepColors={
      Object {
        "PHASE1COMPLETE": "#4EADFF",
        "PHASE1INCOMPLETE": "#B8DEFF",
        "PHASE2COMPLETE": "#F89A1F",
        "PHASE2INCOMPLETE": "#FAB152",
        "PHASE3COMPLETE": "#20B89D",
        "PHASE3INCOMPLETE": "#20DE9D",
      }
    }
    tool={
      Object {
        "content": "Some content",
        "phase": "Some phase",
        "subtitle": "Some subtitle",
        "title": "Some title",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "containerBackgroundColor": undefined,
      "content": "Some content",
      "onPress": [Function],
      "phase": "Some phase",
      "source": "https://images.ctfassets.net/6h184bey8vl3/Rknt0PorAIG4uoMQAg4KM/d697d93618b298ce10d1606c7d3f573a/TimeBug-Icons-_Step-1_.svg",
      "subtitle": "Some subtitle",
      "textStyle": undefined,
      "title": "Some title",
      "tool": Object {
        "content": "Some content",
        "phase": "Some phase",
        "subtitle": "Some subtitle",
        "title": "Some title",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "containerBackgroundColor": undefined,
        "content": "Some content",
        "onPress": [Function],
        "phase": "Some phase",
        "source": "https://images.ctfassets.net/6h184bey8vl3/Rknt0PorAIG4uoMQAg4KM/d697d93618b298ce10d1606c7d3f573a/TimeBug-Icons-_Step-1_.svg",
        "subtitle": "Some subtitle",
        "textStyle": undefined,
        "title": "Some title",
        "tool": Object {
          "content": "Some content",
          "phase": "Some phase",
          "subtitle": "Some subtitle",
          "title": "Some title",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "context": Object {
      "store": Object {
        "dispatch": [Function],
        "getAction": [Function],
        "getActions": [Function],
        "getState": [Function],
        "isActionDispatched": [Function],
        "isActionTypeDispatched": [Function],
        "subscribe": [Function],
      },
    },
    "hydrateIn": undefined,
  },
}
`;
